### DISCUSSION TITLE
Correctly using torchmetrics to measure accuracy

### DISCUSSION THREAD

**Discussion by qlabfgerka** [mrdbourke/pytorch-deep-learning#515] [Q&A]
So in the course videos, accuracy is calculated by writing the custom accuracy function, but how would I correctly implement accuracy from `torchetrics`, or any other metric in that case (for example F1)?

In the code snippet bellow is my (simplified) training function, which calls the `train` step and the `test` step. My question here is, do I need an instance of `torchmetrics.F1Score` for training and testing separately, or is one instance enough?

```py
train_f1_metric = torchmetrics.F1Score(task="multiclass", num_classes=num_classes).to(device)
test_f1_metric = torchmetrics.F1Score(task="multiclass", num_classes=num_classes).to(device)

start_time = timer()

for epoch in tqdm(range(EPOCHS)):
  print(f"Epoch: {epoch}\n")

  epoch_loss, epoch_f1 = train_step(model, train_dataloader, loss_fn, optimizer, train_f1_metric)
  test_step(model, test_dataloader, loss_fn, test_f1_metric)

  print("----------------------------")

end_time = timer()

print(f"\nTotal training time: {end_time - start_time}s")

return ...
```

In the snippet below is my training step function. Here I am updating the f1 using `f1_metric(...)`, I also saw there is an `f1_metric.update(...)` and `f1_metric.forward(...)`, would either of those two be more applicable in this case?

```py
train_loss = 0

model.train()
for (X, y) in dataloader:
  X, y = X.to(device), y.to(device)

  y_pred = model(X)

  loss = loss_fn(y_pred, y)
  train_loss += loss

  f1_metric(y_pred, y)

  optimizer.zero_grad()

  loss.backward()

  optimizer.step()

train_loss /= len(dataloader)
train_f1 = f1_metric.compute()

print(f"Train loss: {train_loss}")
print(f"Train F1: {train_f1}")

f1_metric.reset()

return train_loss, train_f1
```

The problem is a classification problem, using the `torchvision.datasets.CIFAR10` and `torchvision.datasets.CIFAR100` datasets.

So how can I correctly implement the F1 metric (or in general any metric inside of `torchmetrics`, when I figure out the correct implementation, I will measure accuracy, precision and recall next to F1)?

Do I have to have a metric instance separately for a training and testing step?

Does measuring these metrics affect the model learning (so the loss and the time to learn)?

**Comment by abhishekmann** [comment]
Hi @qlabfgerka 

You need to use seperate instances of whatever metric (Accuracy, F1-Score etc.) you are calculating for both train and test sets.

Dr. Sebastian Raschka has a very good post about torchmetrics and will definitely help with your doubt about .forward() and .update()
https://sebastianraschka.com/blog/2022/torchmetrics.html

Hope this helps.
